packages/dev/core/src/Misc/snapshotRenderingHelper.ts

```diff
@@ -0,0 +1,227 @@
+     * @param scene The scene to use the helper in
+     * @param options The options for the helper
+     */
+    constructor(scene: Scene, options?: SnapshotRenderingHelpersOptions) {
+        this._scene = scene;
+        this._engine = scene.getEngine();
+
+        if (!this._engine.isWebGPU) {
+            return;
+        }
+
+        this._options = {
+            morphTargetsNumMaxInfluences: 20,
+            ...options,
+        };
+
+        this._engine.snapshotRenderingMode = Constants.SNAPSHOTRENDERING_FAST;
+
+        this.fixMeshes();
+
+        this._onResizeObserver = this._engine.onResizeObservable.add(() => {
+            const save = this._engine.snapshotRendering;
+            this._engine.snapshotRendering = false;
+            if (save) {
+                this.enableSnapshotRendering();
+            }
+        });
+
+        this._scene.onBeforeRenderObservable.add(() => {
+            if (!this._engine.snapshotRendering || this._engine.snapshotRenderingMode !== Constants.SNAPSHOTRENDERING_FAST) {
+                return;
+            }
+
+            // Animate skeletons
+            scene.skeletons.forEach((skeleton) => skeleton.prepare(true));
+
+            for (const mesh of scene.meshes) {
+                if (mesh.infiniteDistance) {
+                    mesh.transferToEffect(mesh.computeWorldMatrix(true));
+                }
+
+                if (mesh.skeleton) {
+                    mesh.transferToEffect(mesh.computeWorldMatrix(true));
+                }
+
+                if (mesh.morphTargetManager && mesh.subMeshes) {
+                    // Make sure morph target animations work
+                    for (const subMesh of mesh.subMeshes) {
+                        const dw = subMesh._drawWrapper;
+                        const effect = dw.effect;
+                        if (effect) {
+                            const dataBuffer = (dw.drawContext as WebGPUDrawContext).buffers[WebGPUShaderProcessor.LeftOvertUBOName];
+                            const ubLeftOver = (effect._pipelineContext as WebGPUPipelineContext)?.uniformBuffer;
+                            if (dataBuffer && ubLeftOver && ubLeftOver.setDataBuffer(dataBuffer)) {
+                                mesh.morphTargetManager._bind(effect);
+                                BindMorphTargetParameters(mesh, effect);
+                                ubLeftOver.update();
+                            }
+                        }
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * Enable snapshot rendering
+     * Use this method instead of engine.snapshotRendering=true, to make sure everything is ready before enabling snapshot rendering.
+     */
+    public enableSnapshotRendering() {
+        if (!this._engine.isWebGPU) {
+            return;
+        }
+
+        this._scene.executeWhenReady(() => {
+            this._executeAtFrame(this._engine.frameId + 2, () => {

```

Comment:
Just wondering about the arbitrary "2" - it is expected for two frames to end before the flag is set?

Reply:
I want to be sure that a full frame has been executed before the custom function is called, and depending on when this code is executed, this may not be the case. That's why I used “+2” instead of “+1”.